### PR TITLE
feat: add rebalanced shard scheduler

### PR DESCRIPTION
**Pull request by ZuLiangWang**
# Which issue does this PR close?

Closes #

# Rationale for this change
In the current failover implementation, if a node goes down, all its shards will be moved to other shards, and the shards will not be transferred back after the node is restarted, which will cause a lot of pressure on operation and maintenance. We need supports automatic load balancing.

# What changes are included in this PR?
1. Introduced the consistent hash implementation in the golang standard library, ref: https://github.com/golang/groupcache/blob/4a4ac3fbac33b83bb138f808c8945a2812023fc4/consistenthash/consistenthash.go
2. Implemented shard load balancing based on consistent hashing.

# Are there any user-facing changes?
None.

# How does this change test
Pass all unit tests and local tests.

### CODE REVIEW

File: server/coordinator/node_picker.go [hunk 1]
```diff
@@ -22,8 +25,9 @@ func NewRandomNodePicker() NodePicker {
 	return &RandomNodePicker{}
 }
 
-func (p *RandomNodePicker) PickNode(_ context.Context, registeredNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error) {
-	now := time.Now().Unix()
+func (p *RandomNodePicker) PickNode(_ context.Context, _ storage.ShardID, registeredNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error) {
```
Comment (ShiKaiWi) [inline server/coordinator/node_picker.go:30]: The `RandomNodePicker`'s pick logic is totally different from the `ConsistentNodePicker`, which may lead to a mass shard migration after the first shard assignment.

File: server/coordinator/node_picker.go [hunk 2]
```diff
@@ -52,3 +58,33 @@ func (p *RandomNodePicker) PickNode(_ context.Context, registeredNodes []metadat
...
+	for _, registerNode := range registerNodes {
+		if !registerNode.IsExpired(now) {
+			nodeMapping[registerNode.Node.Name] = registerNode
+			hashRing.Add(registerNode.Node.Name)
+		}
+	}
+
+	log.Debug("pick node result", zap.Int("nodeNumber", len(nodeMapping)))
+	if len(nodeMapping) == 0 {
+		return metadata.RegisteredNode{}, errors.WithMessage(ErrNodeNumberNotEnough, "online node length must bigger than 0")
```
Comment (ShiKaiWi) [inline server/coordinator/node_picker.go:84]:
```suggestion
		return metadata.RegisteredNode{}, errors.WithMessage(ErrNodeNumberNotEnough, "at least one online nodes is required")
```

File: server/coordinator/node_picker.go [hunk 3]
```diff
@@ -52,3 +58,33 @@ func (p *RandomNodePicker) PickNode(_ context.Context, registeredNodes []metadat
...
+	}
+
+	pickNodeName := hashRing.Get(strconv.Itoa(int(shardID)))
```
Comment (ShiKaiWi) [inline server/coordinator/node_picker.go:87]:
```suggestion
	pickedNodeName := hashRing.Get(strconv.Itoa(int(shardID)))
```

File: server/cluster/metadata/types.go [hunk 4]
```diff
@@ -136,7 +136,7 @@ func NewRegisteredNode(meta storage.Node, shardInfos []ShardInfo) RegisteredNode
 }
 
 func (n RegisteredNode) IsExpired(now int64) bool {
-	return now >= int64(expiredThreshold)+int64(n.Node.LastTouchTime)
+	return now >= int64(expiredThreshold/time.Millisecond)+int64(n.Node.LastTouchTime)
```
Comment (ShiKaiWi) [inline server/cluster/metadata/types.go:139]: Shall we use the golang `time.Time` to do comparison?

File: server/coordinator/scheduler/scheduler_manager.go [hunk 5]
```diff
@@ -60,7 +60,7 @@ func NewManager(procedureManager procedure.Manager, factory *coordinator.Factory
 		registerSchedulers: []Scheduler{},
 		isRunning:          false,
 		factory:            factory,
-		nodePicker:         coordinator.NewRandomNodePicker(),
+		nodePicker:         coordinator.NewConsistentHashNodePicker(5),
```
[the remark this answers is not in the document]
Comment (ZuLiangWang) [reply server/coordinator/scheduler/scheduler_manager.go:63]: Refer to the configuration of `groupcache` and set its default value to 50.

File: server/coordinator/node_picker.go (hunk 1 above)
Comment (ZuLiangWang) [reply server/coordinator/node_picker.go:30]: It seems that `RandomNodePicker` is unnecessary in the case of `ConsistentHashNodePicker`, just remove it.

File: server/coordinator/node_picker.go [hunk 6]
```diff
@@ -4,51 +4,47 @@ package coordinator
...
-	curOnlineIdx := -1
-	for idx := 0; idx < len(registeredNodes); idx++ {
-		if !registeredNodes[idx].IsExpired(now) {
-			curOnlineIdx++
-		}
-		if curOnlineIdx == selectIdx {
-			return registeredNodes[idx], nil
-		}
-	}
+	pickedNodeName := hashRing.Get(strconv.Itoa(int(shardID)))
```
Comment (ShiKaiWi) [inline server/coordinator/node_picker.go:47]: How about add debug log for the shardID and the pickedNodeName?

File: server/coordinator/node_picker.go [hunk 7]
```diff
@@ -4,51 +4,47 @@ package coordinator
...
-	now := time.Now().Unix()
-	onlineNodeLength := 0
-	for _, registeredNode := range registeredNodes {
-		if !registeredNode.IsExpired(now) {
-			onlineNodeLength++
+func (p *ConsistentHashNodePicker) PickNode(_ context.Context, shardID storage.ShardID, registerNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error) {
+	now := time.Now()
+
+	hashRing := hash.New(p.hashReplicas, nil)
+	nodeMapping := make(map[string]metadata.RegisteredNode, len(registerNodes))
```
Comment (ShiKaiWi) [inline server/coordinator/node_picker.go:34]: I guess there is no need to construct this map? With the hash result, we can find the node just by looping the slice without any memory allocation.

File: server/hash/consistent_hash_ring.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+// This file refer from [groupcache](https://github.com/golang/groupcache/blob/4a4ac3fbac33b83bb138f808c8945a2812023fc4/consistenthash/consistenthash.go)
```
Comment (ShiKaiWi) [inline server/hash/consistent_hash_ring.go:2]:
```suggestion
// The consistent hash ring refers to [groupcache](https://github.com/golang/groupcache/blob/4a4ac3fbac33b83bb138f808c8945a2812023fc4/consistenthash/consistenthash.go)
```

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Pull request merged**